### ISSUE TITLE
[Feature]: Support JSON Schema from llama.cpp

### ISSUE THREAD

**Issue by m0nsky** [SciSharp/LLamaSharp#798]
### Background & Description

I was working on a C# class -> GBNF converter, and came across this:

https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md#json-schemas--gbnf

Here's an example, thanks to [this reddit post](https://www.reddit.com/r/LocalLLaMA/comments/1dj6cpb/comment/l98yute/).

Example JSON:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "reply": {"type": "string"},
    "mood": {"type": "string"},
    "intent": {"type": "string"}
  },
  "required": ["reply", "mood", "intent"]
}
```

Example output:

```
 User: I gave you a cookie.

Please respond in JSON format with your reply, mood, and intent:
{"reply": "Thank you for the cookie! It's delicious!", "mood": "Happy", "intent": "Express gratitude" }
```

**Comment by martindevans** [comment]
Related: https://github.com/SciSharp/LLamaSharp/issues/309 ?

**Comment by m0nsky** [comment]
Yes somewhat, that PR sparked my initial interest in developing the C# class -> GBNF converter. However, the JSON Schema functionality seems to be already implemented on main & server on the llama.cpp side by passing the `-j` parameter, so it wouldn't make sense to reinvent the wheel.

**Comment by RobertRappe** [comment]
As someone slowly working on my own C# agentic system and has been exploring using JSON to format communication with the LLM, getting access to the llama.cpp JSON and grammar tools through LLamaSharp would be really useful.👍

**Comment by phil-scott-78** [comment]
I was working on something with gbnf grammar and c#, but the perf just isn't there. This PR that's soon to be merged is very interesting. Hopefully it'll sneak into the next release, and if so I'll rework my generator to support Lark too

https://github.com/ggerganov/llama.cpp/pull/10224

**Comment by mmoskal** [comment]
@phil-scott-78 for perf with LLGuidance make sure to read up on [lexer](https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md#terminals-vs-rules) ; when you have something working, ping me, we can link your client lib from llguidance repo!

**Comment by phil-scott-78** [comment]
I will certainly reach out. I appreciate that. Looking forward to seeing it merged in and being able to kick the tires

**Comment by phil-scott-78** [comment]
@mmoskal - I will certainly reach out. I appreciate that. Looking forward to seeing it merged in and being able to kick the tires 

Let me ask you a theoretical - the problem we are trying to solve here is json formatting. But that's just the data format we are asking the LLM to generate before deserializing into a .net object

So what if json format wasn't a given? Is there a format that might make better sense than not just llguidance would thrive but the models also could adhere too. Is that a route also worth experimenting?

**Comment by mmoskal** [comment]
@phil-scott-78 TLDR: in theory yes, in practice stick to JSON.

Constrained decoding works best if model most of the time wants to output something compliant with the constraint. If you constraint the output to be an identifier but the model wants to write a Haiku will get something like `silentWavesCrestingUnderMoonlitWinterSky`; or else `sure_I_can_help_you_here_is_your_haiku`. Current models have seen lots of JSON (probably most out of structured data formats, though in some cases Python or JavaScript code might be close second). They were also often fine-tuned to follow a JSON Schema if given in prompt. Thus, unless you're really into training models, I would stick to JSON :)

**Comment by phil-scott-78** [comment]
@mmoskal that's where my head was at, models themselves are gonna wanna drive and use the grammar to keep them between the lines. You mention giving them json schemas, I've had a bunch of luck giving even the smallest of models (like 0.5B with Q4) a sample JSON with templates and a grammar to guide, which is what I was trying to solve with my project.

Anyways, I saw the llama.cpp merge so I gave it a go [updating my project ](https://github.com/phil-scott-78/RedPajama/blob/llguidance/RedPajama/LlGuidanceGrammarGenerator.cs) to output, hopefully, compatible syntaxes. It at least matches the output from your gbnf-to-lark scripts. I was a bit disappointed to see the release bits didn't have the feature enabled, and I'm not gonna lie, I'm afraid I'm not clever enough build it on my windows box. So I'll hold tight until it makes it to llama.cpp and then makes its way here, hopefully.

But progress!

**Comment by jihadkhawaja** [comment]
Thanks @phil-scott-78, Too bad your project only currently supports .NET 9, so I've created my implementation based on [adrienbrault](https://github.com/adrienbrault/json-schema-to-gbnf), it worked flawlessly for my case.

Library/NuGet
[Llama.Grammar](https://github.com/jihadkhawaja/Llama.Grammar)

**Comment by phil-scott-78** [comment]
Love it. I'll check it out when I get a chance

**Comment by phil-scott-78** [comment]
@jihadkhawaja - love the project. I've been dumping my thoughts on the GBNF stuff with .NET into a series on my blog and included a short write up (https://thirty25.blog/blog/2025/05/llama-grammar). Not much more than a rehash of the readme, but I wanted to include it.

**Comment by jihadkhawaja** [comment]
@phil-scott-78 Thanks so much for the write-up and kind words! I really appreciate you taking the time to include Llama.Grammar in your blog.
